tests/tracer/test_processors.py

```diff
@@ -216,3 +217,76 @@ def test_aggregator_partial_flush_2_spans():
     assert writer.pop() == [child1, child2]
     parent.finish()
     assert writer.pop() == [parent]
+
+
+def test_trace_top_level_span_processor_same_service_name():
+    """Parent span and child span have the same service name"""
+    trace_processors = TraceTopLevelSpanProcessor()
+    parent = Span(None, "parent", service="service")
+    parent._local_root = parent
+
+    child1 = Span(None, "child1", service="service")
+    child1.parent_id = parent.span_id
+    child1._parent = parent
+    trace_processors.process_trace([parent, child1])
+    # parent span should be set to top_level
+    # child span should not contain top_level metric
+    assert parent.get_metric("_dd.top_level") == 1
+    assert "_dd.top_level" not in child1.metrics
+
+
+def test_trace_top_level_span_processor_different_service_name():
+    """Parent span and child span have the different service names"""
+    trace_processors = TraceTopLevelSpanProcessor()
+    parent = Span(None, "parent", service="service")
+    parent._local_root = parent
+
+    child2 = Span(None, "child2", service="new_service_name")
+    child2.parent_id = parent.span_id
+    child2._parent = parent
+    trace_processors.process_trace([parent, child2])
+    # child span should be set to top_level
+    assert parent.get_metric("_dd.top_level") == 1
+    assert child2.get_metric("_dd.top_level") == 1
+
+
+def test_trace_top_level_span_processor_orphan_span():
+    """Trace chuck does not contain parent span"""
+    trace_processors = TraceTopLevelSpanProcessor()
+    parent = Span(None, "span_not_in_trace", service="service")
+    parent._local_root = parent
+
+    orphan_span = Span(None, "orphan_span", service="service")
```

Comment:
for this, and all children spans, do we need to be setting `_local_root` explicitly on the spans?

Reply:
Good point. We could but it wouldn't affect the outcome of any of the tests. If `Span._local_root = None` do we assume that it is a local root? I can make this change if the current unit test is confusing

Reply:
I don't think we'd ever end up in a case (using `tracer.trace`) where `Span._local_root = None` right?

in which case how can we be sure we are properly testing realistic scenarios?